Add generic attributes (#26895)
dotnet/docs: docs/csharp/whats-new/csharp-10.md

- [CallerArgumentExpression attribute](#callerargumentexpression-attribute-diagnostics)
- [Enhanced `#line` pragma](#enhanced-line-pragma)

> In addition, there are two features that are available in *preview* mode. You're encouraged to try these features and provide feedback on them. They may change before their final release. In order to use these features, you must [set the `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) in your project.

Reviewer: ```suggestion
In addition, there are two features that are available in *preview* mode. You're encouraged to try these features and provide feedback on them. They may change before their final release. In order to use these features, you must [set `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) in your project.
```

The features available only in preview mode are:

> - [Generic attributes](#generic-attributes)

Reviewer: Is something missing here? Perhaps CallerArgumentExpression?

Author: static abstract interfaces are coming in my next PR. I'll reword this for now.

{
   public TypeAttribute(Type t) => ParamType = t;

   public Type ParamType { get; }
}
```

To apply the attribute, you use the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator:

```csharp
[TypeAttribute(typeof(string))] 
public string Method() => default;
```

Using this feature, you can use a generic attribute instead:

```csharp
public class GenericAttribute<T> : Attribute { }
```

Then, specify the type parameter to use the attribute:

```csharp
[GenericAttribute<string>()]
public string Method() => default;
```

You can apply a fully closed constructed generic attribute. In other words, all type parameters must be specified. For example, the following is not allowed:

```csharp
public class GenericType<T>
{
   [GenericAttribute<T>()] // Not allowed! generic attributes must be fully closed types.
   public string Method() => default;
}
```

The type arguments must satisfy the same restrictions as the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator. Types that require metadata annotations aren't allowed. Examples include the following:

- `dynanmic`

> - `nint`, `nuint`

Author: @RikkiGibson These two work with `typeof`. Should they work here?

Reviewer 2: What happened here is: we took generic attributes as an opportunity to "tighten up" the rules on using type arguments that can't actually be represented in metadata. However, for compat, it proved to be too difficult at least for .NET 6 to bring the same rules back over to typeof, even though typeof has similar limitations as generic attribute type arguments.

Related discussion: dotnet/roslyn#54804

Author: Thanks @RikkiGibson That discussion helped. I added some clarifications and what types to use instead.

- `object?` (or any nullable reference type)

> - `(int X, int Y)` (or any other tuple types using C# tuple syntax). Note that you can use the underlying type of `ValueTuple<int, int>`.

Author: @RikkiGibson This also works with `typeof`. Same question.

## Enhanced #line pragma

C# 10 supports a new format for the `#line` pragma. You likely won't use the new format, but you'll see its effects. The enhancements enable more fine-grained output in domain-specific languages (DSLs) like Razor. The Razor engine uses these enhancements to improve the debugging experience. You'll find debuggers can highlight your Razor source more accurately. To learn more about the new syntax, see the article on [Preprocessor directives](../language-reference/preprocessor-directives.md#error-and-warning-information) in the language reference. You can also read the [feature specification](~/_csharplang/proposals/csharp-10.0/enhanced-line-directives.md#examples) for Razor based examples.

## Generic attributes

> [!IMPORTANT]

> > *Generic attributes* is a preview feature. You must [set the `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) to enable this feature. This feature may change before its final release.

Reviewer: ```suggestion
> *Generic attributes* is a preview feature. You must [set `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) to enable this feature. This feature may change before its final release.
```

> You can declare a [generic class](../programming-guide/generics/generic-classes.md) whose base class is <xref:System.Attribute?displayProperty=fullName>. This provides a more convenient syntax for atributes that require a <xref:System.Type?displayProperty=nameWithType> parameter. Previously, you'd need to write an attribute that takes a `Type` as its parameter:

Reviewer: ```suggestion
You can declare a [generic class](../programming-guide/generics/generic-classes.md) whose base class is <xref:System.Attribute?displayProperty=fullName>. This provides a more convenient syntax for attributes that require a <xref:System.Type?displayProperty=nameWithType> parameter. Previously, you'd need to create an attribute that takes a `Type` as its constructor parameter:
```

```csharp
public class TypeAttribute : Attribute
{
   public TypeAttribute(Type t) => ParamType = t;

   public Type ParamType { get; }
}
```

> To apply the attribute, you use the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator:

Reviewer: ```suggestion
And to apply the attribute, you use the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator:
```

```csharp
[TypeAttribute(typeof(string))] 
public string Method() => default;
```

> Using this feature, you can use a generic attribute instead:

Reviewer: ```suggestion
Using this new feature, you can create a generic attribute instead:
```

```csharp
public class GenericAttribute<T> : Attribute { }
```

Then, specify the type parameter to use the attribute:

```csharp
[GenericAttribute<string>()]
public string Method() => default;
```

You can apply a fully closed constructed generic attribute. In other words, all type parameters must be specified. For example, the following is not allowed:

```csharp
public class GenericType<T>
{
   [GenericAttribute<T>()] // Not allowed! generic attributes must be fully closed types.
   public string Method() => default;
}
```

The type arguments must satisfy the same restrictions as the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator. Types that require metadata annotations aren't allowed. Examples include the following:

- `dynamic`
- `nint`, `nuint`
- `string?` (or any nullable reference type)
- `(int X, int Y)` (or any other tuple types using C# tuple syntax).

These types aren't directly represented in metadata. They types include annotations that describe the type. In all cases, you can use the underlying type instead:

- `object` for `dynamic`.

> - <xref:System.IntPtr> instead of `nint`, or `unint`.

Reviewer: ```suggestion
- <xref:System.IntPtr> instead of `nint` or `unint`.
```
